### ISSUE TITLE
sbt 0.13.6 not compatible with sbt-groll built against 0.13.5

### ISSUE THREAD

**Issue by hseeberger** [sbt/sbt#1599]
Use this in `project/plugins`:

```
addSbtPlugin("name.heikoseeberger" % "sbt-groll" % "3.1.4")
```

Then start sbt and enter `groll list`. This will lead to:

```
test> groll list
scala.MatchError: ThisProject (of class sbt.ThisProject$)
    at sbt.Scope$.resolveProjectRef(Scope.scala:89)
    at sbt.Scope$.resolveReference(Scope.scala:85)
    at sbt.Load$$anonfun$defaultDelegates$1.sbt$Load$$anonfun$$resolveRef$1(Load.scala:95)
    at sbt.Load$$anonfun$defaultDelegates$1$$anonfun$apply$3.apply(Load.scala:99)
    at sbt.Load$$anonfun$defaultDelegates$1$$anonfun$apply$3.apply(Load.scala:99)
    at sbt.Scope$.indexedDelegates(Scope.scala:197)
    at sbt.Scope$$anonfun$delegates$1.apply(Scope.scala:173)
    at sbt.Scope$$anonfun$delegates$1.apply(Scope.scala:173)
    at sbt.Settings0.get(Settings.scala:25)
    at sbt.Scoped$DefinableSetting$class.get(Structure.scala:151)
    at sbt.SettingKey.get(Structure.scala:34)
    at name.heikoseeberger.sbtgroll.package$.setting(package.scala:62)
    at name.heikoseeberger.sbtgroll.Groll.<init>(Groll.scala:33)
    at name.heikoseeberger.sbtgroll.Groll$.apply(Groll.scala:26)
    at name.heikoseeberger.sbtgroll.SbtGroll$$anonfun$name$heikoseeberger$sbtgroll$SbtGroll$$grollCommand$2.apply(SbtGroll.scala:61)
    at name.heikoseeberger.sbtgroll.SbtGroll$$anonfun$name$heikoseeberger$sbtgroll$SbtGroll$$grollCommand$2.apply(SbtGroll.scala:61)
    at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:58)
    at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:58)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:60)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:60)
    at sbt.Command$.process(Command.scala:92)
    at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
    at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
    at sbt.State$$anon$1.process(State.scala:184)
    at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
    at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.MainLoop$.next(MainLoop.scala:98)
    at sbt.MainLoop$.run(MainLoop.scala:91)
    at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:70)
    at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:65)
    at sbt.Using.apply(Using.scala:24)
    at sbt.MainLoop$.runWithNewLog(MainLoop.scala:65)
    at sbt.MainLoop$.runAndClearLast(MainLoop.scala:48)
    at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:32)
    at sbt.MainLoop$.runLogged(MainLoop.scala:24)
    at sbt.StandardMain$.runManaged(Main.scala:53)
    at sbt.xMain.run(Main.scala:28)
    at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
    at xsbt.boot.Launch$.withContextLoader(Launch.scala:128)
    at xsbt.boot.Launch$.run(Launch.scala:109)
    at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:35)
    at xsbt.boot.Launch$.launch(Launch.scala:117)
    at xsbt.boot.Launch$.apply(Launch.scala:18)
    at xsbt.boot.Boot$.runImpl(Boot.scala:41)
    at xsbt.boot.Boot$.main(Boot.scala:17)
    at xsbt.boot.Boot.main(Boot.scala)
```

**Comment by hseeberger** [comment]
It looks like commit `adb41611` is to blame that `ThisProject` is no longer matched:

```
   def resolveProjectRef(current: URI, rootProject: URI => String, ref: ProjectReference): ProjectRef =
     ref match {
-      case ThisProject | LocalRootProject => ProjectRef(current, rootProject(current))
-      case LocalProject(id)               => ProjectRef(current, id)
-      case RootProject(uri)               =>
+      case LocalRootProject => ProjectRef(current, rootProject(current))
+      case LocalProject(id) => ProjectRef(current, id)
+      case RootProject(uri) =>
         val res = resolveBuild(current, uri); ProjectRef(res, rootProject(res))
-      case ProjectRef(uri, id)            => ProjectRef(resolveBuild(current, uri), id)
+      case ProjectRef(uri, id) => ProjectRef(resolveBuild(current, uri), id)
     }
```

**Comment by jsuereth** [comment]
@hseeberger interesting, because that was a scalariform style fix, nothing more.   I'll add a quick patch, but two questions:
1. Do you have a mechanism to reproduce the ThisProject issue without groll?
2. Were you aware of scalariform volatile rewrites?   This discovery makes me a bit nervous now that perhaps we lost other bits of necessary functionality.

cc> @eed3si9n

**Comment by hseeberger** [comment]
Look at the above diff. Clearly not a formatting thingy, but there's no more matching for `ThisProject`.

Regarding (1) I haven't narrowed it down, but it should be fairly simple, just use `ThisProject` in your build.

**Comment by jsuereth** [comment]
@hseeberger The commit you point it was ONLY running scalariform and committing.  My guess is there was a bug with:

``` scala
  case ThisProject | LocalRootProject =>
```

where it drops part of the pattern.

**Comment by hseeberger** [comment]
@jsuereth Ah, got you. Scary!

**Comment by retronym** [comment]
I would recommend a bytecode diff before and after that commit.

**Comment by jsuereth** [comment]
@retronym good call, we'll set that up.

**Comment by retronym** [comment]
FWIW, I can't reproduce this in isolation.

```
% tail -n100 project/plugins.sbt build.sbt src/main/scala/test.scala
==> project/plugins.sbt <==
addSbtPlugin("com.typesafe.sbt" % "sbt-scalariform" % "1.3.0")

==> build.sbt <==
scalaVersion := "2.11.2"

libraryDependencies += "org.scalacheck" %% "scalacheck" % "1.11.5"

scalariformSettings

{
  import com.typesafe.sbt.SbtScalariform.{ ScalariformKeys => sr, _ }
  import scalariform.formatter.preferences._
  Seq(
    sr.preferences := sr.preferences.value.setPreference(AlignSingleLineCaseStatements, true)
  )
}

==> src/main/scala/test.scala <==
class     C {

  def resolveProjectRef(current: URI, rootProject: URI => String, ref: ProjectReference): ProjectRef =
    ref match {
      case ThisProject | LocalRootProject => ProjectRef(current, rootProject(current))
      case LocalProject(id)               => ProjectRef(current, id)
      case RootProject(uri) =>
        val res = resolveBuild(current, uri); ProjectRef(res, rootProject(res))
      case ProjectRef(uri, id) => ProjectRef(resolveBuild(current, uri), id)
    }
}

% sbt compile:scalariformFormat
[info] Loading global plugins from /Users/jason/.sbt/0.13/plugins
[info] Loading project definition from /Users/jason/code/scratch3/project
[info] Set current project to scratch3 (in build file:/Users/jason/code/scratch3/)
[info] Formatting 1 Scala source {file:/Users/jason/code/scratch3/}scratch3(compile) ...
[info] Reformatted 1 Scala source {file:/Users/jason/code/scratch3/}scratch3(compile).
[success] Total time: 0 s, completed Sep 15, 2014 11:00:31 PM

% cat src/main/scala/test.scala
class C {

  def resolveProjectRef(current: URI, rootProject: URI => String, ref: ProjectReference): ProjectRef =
    ref match {
      case ThisProject | LocalRootProject => ProjectRef(current, rootProject(current))
      case LocalProject(id)               => ProjectRef(current, id)
      case RootProject(uri) =>
        val res = resolveBuild(current, uri); ProjectRef(res, rootProject(res))
      case ProjectRef(uri, id) => ProjectRef(resolveBuild(current, uri), id)
    }
}
```

**Comment by jsuereth** [comment]
@retronym Yeah, as I investigate the issue, i'm finding all notion of ThisProject in scope disappeared.  I'm suspecting something else at play than just the commit reported.

**Comment by jsuereth** [comment]
Ok, @hseeberger this is actually as designed.  The commit your reference was incorrect, here's your offending comit: https://github.com/sbt/sbt/commit/5deb103ef6

Now, I agree with the sentiment in that commit. Can i ask your use case of hitting resolveProjectRef directly?  It could be we have a hole here somewhere with the API, and I'd rather move you somewhere a bit more stable.

**Comment by hseeberger** [comment]
I'm using 

```
key in ThisProject get structure(state).data getOrElse sys.error(s"$key undefined!")
```

**Comment by eed3si9n** [comment]
> Ok, @hseeberger this is actually as designed. The commit your reference was incorrect, here's your offending comit: 5deb103

O man. You guys scared me there. Thank goodness scalariform is correct.

**Comment by jsuereth** [comment]
@hseeberger Ah, this API, I consider "advanced".  If you have extracted ( I see structure) can you modify your code to:

``` scala
key in extracted.currentRef get structure(state).data getOrElse sys.error(s"$key undefined!")
```

The reality is, the resolution of `ThisProject` you got before was _wrong_, even if it generally worked ok.  It would ALWAYS point at the _root_ project even if someone used the `project <name>` command in sbt to set a different current project (i.e. ThisProject should mean something else).

Also, you're using advanced sbt API.   ThisProject is not safe to use external to `Setting[_]` instances.  i.e. it's safe for users, but not for internal API calls (Commands, state, etc.).

**Comment by hseeberger** [comment]
Thanks for this lesson ;-)
Will give it a try ...

**Comment by hseeberger** [comment]
@jsuereth, works like a charm. Feel free to close this issue.

**Comment by hseeberger** [comment]
@jsuereth, thanks a lot for the quick help!
Just released sbt-groll 3.2.0 with this change.

**Comment by viktorklang** [comment]
What rule would rewrite that?
